File: api/app/domains/listings/dto.py
from dataclasses import dataclass
from datetime import datetime
from decimal import Decimal
from uuid import UUID


@dataclass(frozen=True, slots=True)
class ListingImageResult:
    id: UUID
    url: str
    content_type: str
    position: int
    created_at: datetime
    updated_at: datetime


@dataclass(frozen=True, slots=True)
class ListingResult:
    id: UUID
    slug: str
    name: str
    description: str
    price_per_night: Decimal
    max_guests: int
    created_at: datetime
    updated_at: datetime
    images: list[ListingImageResult]

File: web/src/entities/listing/api/dto.ts
import { z } from "zod";

export const listingImageDtoSchema = z.object({
  id: z.uuid(),
  url: z.url(),
  contentType: z.string(),
  position: z.number().int(),
  createdAt: z.iso.datetime(),
  updatedAt: z.iso.datetime(),
});

export const listingDtoSchema = z.object({
  id: z.uuid(),
  slug: z.string(),
  name: z.string(),
  description: z.string(),
  pricePerNight: z.string(),
  maxGuests: z.number().int(),
  createdAt: z.iso.datetime(),
  updatedAt: z.iso.datetime(),
  images: z.array(listingImageDtoSchema),
});

export type ListingDto = z.infer<typeof listingDtoSchema>;
